ipq40xx: Add support for TP-Link Deco M5

DropDemBits opened the pull request:
Specifications
--------------
- SoC          : Qualcomm IPQ4019
- RAM          : 256 MiB DDR3-1600 (NT5CC128M16IP)
- Flash        : 32 MiB SPI NOR
  - V1, V2     : GigaDevice GD25Q256C
  - V3         : Winbond W25Q256FV, Micron N25Q128A11
  - V320       : EON EN25QH256, XMC XM25QH256BK
- WLAN         : IPQ4019 On-chip 
  - 2.4 GHz    : 2x2 MIMO 802.11b/g/n
  - 5 GHz      : 2x2 MIMO 802.11n/ac
- Ethernet     : QCA8072 10/100/1000BASE-T 1x WAN; 1x LAN
- UART         : 3v3 115200n, use pins under heatsink
- Buttons      : 1x Reset
- LEDs         : 1x Combined RGB LED
  - V1         : Driven by LP5521 channels (Channel 0, 1, 2?)
  - V3, V320   : Driven by GPIO Pins (Pin 28, 32, 35 Active Low)
- Bluetooth    :
  - V1, V2, V3 : CSR8811
  - V320       : AC6368A/B
- Power        : DC 12V @ 1.2A
- FCC ID       :
  - V1         : TE7M5
  - V3         : TE7M5V3
  - V320       : TE7M5V32, 2AXJ4M5V3, 2BCGWM5V3
- TFTP ID      :
  - Client     : 192.168.0.66
  - Router     : 192.168.0.11

MAC Addresses
-------------
Label   : OPAQUE partition @ 0x0008
LAN     : Label
WAN     : Label + 1
WLAN 2G : Label + 2
WLAN 5G : Label + 3

Installation
------------
1. Install TFTP server software
2. Rename the factory image to 'M5v1_tp_recovery.bin' and copy it into
   the TFTP folder
3. Connect an interface configured to 192.168.0.66/24 to the LAN port
   - Note: The LAN port is the ethernet port that is furthest away from
     the power plug
4. Press the 'reset' button down before powering up the device, holding
   for at least 10 seconds
5. Router should download the factory image, install it, and then reboot
6. Router will start up with address 192.168.1.1


Known Issues
------------
- ~~LEDs configuration on V1 and V2 does not work~~ Fixed by enabling LP5521 driver
- Devices with GD25Q256C flash might not be detected by SFDP probing on Linux 6.12

DropDemBits committed:
ipq40xx: add support for TP-Link Deco-M5

Specifications
--------------
- SoC          : Qualcomm IPQ4019
- RAM          : 256 MiB DDR3-1600 (NT5CC128M16IP)
- Flash        : 32 MiB SPI NOR
  - V1, V2     : GigaDevice GD25Q256C
  - V3         : Winbond W25Q256FV, Micron N25Q128A11
  - V320       : EON EN25QH256, XMC XM25QH256BK (with patch included)
- WLAN         : IPQ4019 On-chip
  - 2.4 GHz    : 2x2 MIMO 802.11b/g/n
  - 5 GHz      : 2x2 MIMO 802.11n/ac
- Ethernet     : QCA8072 10/100/1000BASE-T 1x WAN; 1x LAN
- UART         : 3v3 115200n, use pins under heatsink
- Buttons      : 1x Reset
- LEDs         : 1x Combined RGB LED
  - V1         : Driven by LP5521 channels (Channel 0, 1, 2?)
  - V3, V320   : Driven by GPIO Pins (Pin 28, 32, 35 Active Low)
- Bluetooth    :
  - V1, V2, V3 : CSR8811
  - V320       : AC6368A/B
- Power        : DC 12V @ 1.2A
- FCC ID       :
  - V1         : TE7M5
  - V3         : TE7M5V3
  - V320       : TE7M5V32, 2AXJ4M5V3, 2BCGWM5V3
- TFTP ID      :
  - Client     : 192.168.0.66
  - Router     : 192.168.0.11

MAC Addresses
-------------
Label   : OPAQUE partition @ 0x0008
LAN     : Label
WAN     : Label + 1
WLAN 2G : Label + 2
WLAN 5G : Label + 3

Installation
------------
1. Install TFTP server software
2. Rename the factory image to 'M5v1_tp_recovery.bin' and copy it into
   the TFTP folder
3. Connect an interface configured to 192.168.0.66/24 to the LAN port
   - Note: The LAN port is the ethernet port that is furthest away from
     the power plug
4. Press the 'reset' button down before powering up the device, holding
   for at least 10 seconds
5. Router should download the factory image, install it, and then reboot
6. Router will start up with address 192.168.1.1

Known Issues
------------
- Some devices (including some V1 and V2 devices) may have flash chips
  that are not detected, supported by, or have broken SFDP support

Signed-off-by: Bee Cadorna <r3usrlnd@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17537
Signed-off-by: Robert Marko <robimarko@gmail.com>

DropDemBits commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-deco-m5.dtsi:
```diff
@@ -0,0 +1,378 @@
+
+/ {
+	chosen {
+		bootargs-append = " root=/dev/mtdblock16";
```
I've not been able to figure out why OpenWrt isn't finding the rootfs partition, because even though the bootloader passes in `root=mtd:1:rootfs rootfstype=squashfs rootwait` and fails to find `mtd:1:rootfs`, it should fall back to finding the rootfs by partition name after the `firmware` partition is split?

DropDemBits replied:
I suspect that it's something to do with the stock U-Boot version and it potentially not supporting partition splitting. We could split `firmware` into `1:hlos` and `1:rootfs` like what the OEM firmware does, though we'd lose out on flexible kernel & rootfs partitions.

robimarko commented:
SPI-NOR support should go upstream, but do note since it supports SFDP that unless its broken it should work with the generic driver

DropDemBits committed:
ipq40xx: enable LP5521 driver for TP-Link Deco M5

TP-Link Deco M5 v1 and v2 models use a separate LP5521 controller to
control the LEDs on the device.

Signed-off-by: Bee Cadorna <r3usrlnd@gmail.com>
Tested-by: Tan Li Boon <undisputed.seraphim@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17537
Signed-off-by: Robert Marko <robimarko@gmail.com>

DropDemBits commented:
Current status: Waiting for testing verification for
- [ ] Enabling LP5521 configuration to fix v1 & v2 LED functionality (see https://forum.openwrt.org/t/ipq4019-adding-support-for-tp-link-deco-m5/85061/431) 
- [ ] Removing XMC patch for v3.20 and select v3 models to delegate to the generic SPI-NOR functionality (see https://forum.openwrt.org/t/ipq4019-adding-support-for-tp-link-deco-m5/85061/432)

DropDemBits commented:
@robimarko
Verified that the SPI-NOR patch is no longer needed, see https://forum.openwrt.org/t/ipq4019-adding-support-for-tp-link-deco-m5/85061/440

robimarko commented on package/firmware/ipq-wifi/board-tplink_deco-m5.qca4019:
This goes into the qca-wireless repo, not here locally

DropDemBits replied:
Will delete file once https://github.com/openwrt/firmware_qca-wireless/pull/59 is merged.

robimarko replied:
I merged the BDF

aarnaud replied:
so this commit should be remove in this PR https://github.com/openwrt/openwrt/pull/17537/commits/0ecf04be60445f9bfbf54119aab79574bd77977e and we should be able to merge right ?

robimarko replied:
Unlikely, there is probably more stuff to fixup

aarnaud commented:
thanks @DropDemBits  for the PR and @robimarko for the review, Is there any change to do ? or know issue ? I have a Deco M5 with UART and u-boot access + original FW dump. I would like to use openwrt on it.

robimarko commented:
@DropDemBits Whats the status?

aarnaud commented:
> thanks @DropDemBits for the PR and @robimarko for the review, Is there any change to do ? or know issue ? I have a Deco M5 with UART and u-boot access + original FW dump. I would like to use openwrt on it.

I did compiled for Deco M5 v3, it's working fine. I'm thankful. Don't hesitate if you need support to test things.

I hope this can be merge soon.

undisputed-seraphim commented:
https://github.com/openwrt/firmware_qca-wireless/pull/59 has been merged. Bumping for more reviews to this MR. Very thankful for the hard work @DropDemBits and co have put in; I'm ready to help get this merged.

DropDemBits commented:
@robimarko

Current status:

- Dropped temporary BDF commit, and rebased PR to latest master.
- @dutchmillbytes has reported that the SFDP detection for GD25Q256C-based devices (V1 & V2) may not work, though I'm not sure on the specifics of the issue.

While there are more hardware sub-versions (V2.10, other V3.XX) here than are listed, I believe that they differ by just flash chips and/or BOM, rather than anything requiring a separate DTS file.

dutchmillbytes commented:
The presence of the `PARSE_SFDP` flag leads to a failure of detecting the GigaDevice GD25Q256C. Subsequently the rootfs is not found and booting fails. This [patch](https://github.com/dutchmillbytes/openwrt/blob/c666fd5bbbe483ec2844326110951b76b9c80f25/target/linux/ipq40xx/patches-6.6/450-fix-detection-gigadevice-deco-m5.patch) has eliminated the issue.

@DropDemBits  Can you add this patch to your PR?

A bit of nitpicking: there is a small typo in `target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-deco-m5-v2.dts`: the `ath10k-calibration-variant` entry refers to `v1` instead of `v2`.

The recent compiled firmware for 24.10.1 (based on this PR and the kernel patch) has been verified to be working on v1, v2 and v3 devices.

DropDemBits commented:
@dutchmillbytes

> The presence of the `PARSE_SFDP` flag leads to a failure of detecting the GigaDevice GD25Q256C.

Hmm, this implies that the GD25Q256C's SFDP is wrong in some way.
However, the patch as it is right now also disables SFDP parsing for the GD25Q256D and GD25Q256E, even though those appear to behave properly according to <https://lore.kernel.org/all/20211207140254.87681-14-tudor.ambarus@microchip.com/>. I'll try to see if I can come up with a more targeted patch in the coming days, but no guarantee.

It'd also be nice if someone is able to dump the SFDP for the GD25Q256C, either through using the SPI-NOR legacy fallback path or dumping it directly from the chip, so that it can be cross-referenced with the datasheet.

@robimarko I presume whatever the final patch ends up being would need to be sent upstream instead of living in the OpenWrt repo, correct? I imagine the current "disable SFDP" patch wouldn't be accepted as it also affects the GD25Q256D and GD25Q256E.

> the `ath10k-calibration-variant` entry refers to `v1` instead of `v2`.

AFAIK, the v1 and v2 do not have distinct FCC ID's, which implies that they use the same antenna arrangement and thus the same calibration data. We could duplicate the board file to have a distinct v2 version, but it's better storage-space wise to just reuse the same calibration variant.

survivor182ca commented:
Go go go, it has to be merge before this device get obsolete. Try many version of this code and it works well on my deco m5 v3

survivor182ca commented:
If it doesn t work with 1% of the device, the community will fix it later once it is merge to master (bigger community). Also, the worse that can happen is tftp recovery for which it doesn t involve any soldering.

DropDemBits commented:
@russinnes Didn't know about alt names, thanks! Though, it is a little bit annoying that the alt variants are displayed before the main variant (i.e. v2 ends up being listed before v1).

aarnaud commented:
Hi here, what is missing about this PR about Deco M5 ? it will be great to use the master branch to be able to build it. 
Can I help ? I have one Deco M5 v3 (US) with UART pin soldered

robimarko commented:
Needs to be rebased as ipq40xx uses 6.12 now instead of 6.6

hauke commented:
There is a compile problem in one of the dts files:
```
make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-arm-openwrt-linux-muslgnueabi_musl/linux-ipq40xx_generic/linux-6.12.55'
make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
cc1: fatal error: /__w/openwrt/openwrt/openwrt/build_dir/target-arm-openwrt-linux-muslgnueabi_musl/linux-ipq40xx_generic/linux-6.12.55/arch/arm/boot/dts/qcom/qcom-ipq4019-deco-m5-v1.dts: No such file or directory
compilation terminated.
make[4]: *** [Makefile:24: /__w/openwrt/openwrt/openwrt/build_dir/target-arm-openwrt-linux-muslgnueabi_musl/linux-ipq40xx_generic/image-qcom-ipq4019-deco-m5-v1.dtb] Error 1
make[3]: *** [Makefile:24: dtb] Error 2
make[2]: *** [Makefile:12: dtb] Error 2
```

achterin commented:
looks like the dts files are still in the kernel-6.6 directory and need to be moved.

DropDemBits commented:
Moved the dts(i) files over (oops!) and rebased against main

robimarko commented:
I am not the biggest fan of backporting devices to a stable branch (Like what is the point of a stable branch).
But since this is just DTS plus LED driver its fine, just make sure to put the DTS in the old location

achterin commented:
Thanks for finally bringing support for this device family to OpenWrt! I'd also like to see these devices as a backport to 25.12, so they can be used in Gluon/Freifunk soon.
